### PR TITLE
Reuse the InputEntityReader in SettableByteEntityReader

### PR DESCRIPTION
**Pull request by abhishekagarwal87**
Using `InputFormat` instead of `ParseSpec` can be very slow if complex flattening scheme is used. This is because with InputFormat, we create an `InputEntityReader` for every stream chunk. It was assumed that creating the reader is a cheap operation but it is not in some cases. 

Here are two flame graphs to illustrate the difference
Task runner thread with InputFormat
Screenshot 2022-02-18 at 1 42 58 PM

Task runner thread with ParseSpec
Screenshot 2022-02-18 at 1 41 46 PM

Instead of fixing the JsonParser, I have added a new class called `SettableByteEntity` that allows one to add new binary data without creating a new InputStream. So we can create an InputEntityReader just once instead of creating it for every stream chunk. Readers for different formats consume one chunk at a time from input stream while the StreamChunkParser sets the new chunk on the input stream. This all happens in a single thread. 

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by jihoonson** [comment]
@abhishekagarwal87 thanks for making this PR. The overall idea SGTM. Can you check the Travis failure? Two indexing module test jobs timed out while running kinesis ingestion tests which seems suspicious.

**Comment by abhishekagarwal87** [comment]
@jihoonson - Yes. The failures are definitely related. I will address them.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SettableByteEntity.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+      if (null != delegate && available() > 0) {
+        throw new IAE("New data cannot be set in buffer till all the old data has been read");
+      }
+      this.delegate = new ByteBufferInputStream(newBuffer);
+    }
+
+    @Override
+    public int read()
+    {
+      Preconditions.checkNotNull(delegate, "Buffet is not set");
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SettableByteEntity.java:106]: Nit:
```suggestion
      Preconditions.checkNotNull(delegate, "Buffer is not set");
```

Hope, it wasn't on purpose! 😂

**Review by kfaraz — APPROVED** [review]
LGTM. Minor nitpicks.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SettableByteEntity.java (hunk 1 above)
Comment (abhishekagarwal87) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SettableByteEntity.java:106]: lol. I should fix this.

**Comment by abhishekagarwal87** [comment]
Thank you @kfaraz. I have addressed your comments.

**Review by jihoonson — APPROVED** [review]
LGTM

**Pull request merged**